File: Reference/src/Brontide.Reference.Experimental.Binding/Portable/PortableEnvelope.cs
using System.Collections.Immutable;

namespace Brontide.Reference.Experimental.Binding.Portable;

/// <summary>The Channel envelope kinds.</summary>
/// <remarks>
/// There is deliberately no denial kind. A local authority denial emits no frame at all, so the
/// Cooling experiment's denial message is not carried into the portable envelope set. Process loss
/// is likewise observed locally rather than signalled.
/// </remarks>
public enum PortableEnvelopeKind
{
    Establish,
    EstablishAccepted,
    Ready,
    Request,
    Outcome,
    ProtocolError,
    Withdraw,
    Terminate
}

public static class PortableEnvelopeKinds
{
    public static string Token(this PortableEnvelopeKind kind) => kind switch
    {
        PortableEnvelopeKind.Establish => "establish",
        PortableEnvelopeKind.EstablishAccepted => "establish-accepted",
        PortableEnvelopeKind.Ready => "ready",
        PortableEnvelopeKind.Request => "request",
        PortableEnvelopeKind.Outcome => "outcome",
        PortableEnvelopeKind.ProtocolError => "protocol-error",
        PortableEnvelopeKind.Withdraw => "withdraw",
        PortableEnvelopeKind.Terminate => "terminate",
        _ => throw new ArgumentOutOfRangeException(nameof(kind))
    };

    public static PortableEnvelopeKind Parse(string token)
    {
        foreach (var kind in Enum.GetValues<PortableEnvelopeKind>())
        {
            if (string.Equals(kind.Token(), token, StringComparison.Ordinal))
            {
                return kind;
            }
        }

        throw new PortableFaultException(
            PortableProtocolCategory.UnsupportedKind,
            "unknown-kind",
            $"Envelope kind '{token}' is outside the declared set; an unknown kind is never treated as a no-op.");
    }

    /// <summary>A request, an outcome, and a correlated protocol error all require a request identity.</summary>
    public static bool RequiresRequestId(this PortableEnvelopeKind kind) =>
        kind is PortableEnvelopeKind.Request or PortableEnvelopeKind.Outcome;
}

/// <summary>One Channel envelope.</summary>
public sealed record PortableEnvelope(
    int ContractVersion,
    PortableEnvelopeKind Kind,
    PortableChannelId ChannelId,
    PortableChannelRequestId? RequestId,
    PortableChannelExecutionId? ExecutionId,
    CborItem Body)
{
    public static PortableEnvelope Control(
        PortableEnvelopeKind kind,
        PortableChannelId channel,
        CborItem? body = null) =>
        new(
            PortableContractDocument.SupportedContractVersion,
            kind,
            channel,
            null,
            null,
            body ?? CborMap.Empty);
}

/// <summary>Strict envelope codec; every miss is a portable category rather than a parser failure.</summary>
public static class PortableEnvelopeCodec
{
    private static readonly string[] EnvelopeFields =
    [
        "contractVersion", "kind", "channelId", "requestId", "executionId", "body"
    ];

    public static byte[] Encode(PortableEnvelope envelope)
    {
        ArgumentNullException.ThrowIfNull(envelope);
        var entries = new List<(string, CborItem)>
        {
            ("contractVersion", new CborInteger(envelope.ContractVersion)),
            ("kind", new CborText(envelope.Kind.Token())),
            ("channelId", new CborText(envelope.ChannelId.Value)),
            ("body", envelope.Body)
        };

        if (envelope.RequestId is { } request)
        {
            entries.Add(("requestId", new CborText(request.Value)));
        }

        if (envelope.ExecutionId is { } execution)
        {
            entries.Add(("executionId", new CborText(execution.Value)));
        }

        return PortableCbor.Encode(CborMap.Of(entries));
    }

    public static PortableEnvelope Decode(ReadOnlySpan<byte> body, PortableLimits limits)
    {
        var item = PortableCbor.Decode(body, limits);
        var map = CborAccess.RequireMap(item, "envelope");

        // Foreign runtime identity is refused before the field set is checked, so the rejection
        // names the rule that actually applies rather than reporting an unknown field.
        foreach (var entry in map.Entries.Where(entry => !string.Equals(entry.Key, "body", StringComparison.Ordinal)))
        {
            if (PortableForbiddenContent.IsForbidden(entry.Key))
            {
                throw PortableFaultException.Malformed(
                    "foreign-runtime-data",
                    $"Control field '{entry.Key}' carries foreign runtime identity.");
            }

            PortableForbiddenContent.RequireCleanControlItem(entry.Value);
        }

        map.RequireDeclaredFields("envelope", EnvelopeFields);

        var contractVersion = map.RequireInt32("contractVersion");
        if (contractVersion != PortableContractDocument.SupportedContractVersion)
        {
            throw new PortableFaultException(
                PortableProtocolCategory.UnsupportedVersion,
                "envelope-version",
                $"Envelope contract version {contractVersion} is not recognized.");
        }

        var kind = PortableEnvelopeKinds.Parse(map.RequireText("kind"));
        var channelId = new PortableChannelId(map.RequireText("channelId"));
        var requestText = map.OptionalText("requestId");
        var executionText = map.OptionalText("executionId");

        if (kind.RequiresRequestId() && requestText is null)
        {
            throw PortableFaultException.Malformed(
                "correlation-absent",
                $"A '{kind.Token()}' envelope carries a request identity; it is never matched to an outstanding request by position.");
        }

        return new PortableEnvelope(
            contractVersion,
            kind,
            channelId,
            requestText is null ? null : new PortableChannelRequestId(requestText),
            executionText is null ? null : new PortableChannelExecutionId(executionText),
            map.Require("body"));
    }
}

/// <summary>The body of an <c>establish-accepted</c> envelope.</summary>
public sealed record PortableEstablishAcceptedBody(
    PortableContractDocument Contract,
    ImmutableArray<PortableCompactAssignment> CompactIdentifiers)
{
    public CborItem Encode() => CborMap.Of(
    [
        ("contract", PortableContractCodec.Encode(Contract)),
        ("compactIdentifiers", new CborArray(
        [
            .. CompactIdentifiers.Select(assignment => (CborItem)CborMap.Of(
            [
                ("space", new CborText(assignment.Space.ToString().ToLowerInvariant())),
                ("reference", new CborText(assignment.Reference)),
                ("compactId", new CborInteger(assignment.CompactId.Value))
            ]))
        ]))
    ]);

    public static PortableEstablishAcceptedBody Decode(CborItem item)
    {
        var map = CborAccess.RequireMap(item, "establish-accepted");
        map.RequireDeclaredFields("establish-accepted", "contract", "compactIdentifiers");
        return new PortableEstablishAcceptedBody(
            PortableContractCodec.Decode(map.Require("contract")),
            map.RequireArray("compactIdentifiers", element =>
            {
                var entry = CborAccess.RequireMap(element, "compactAssignment");
                entry.RequireDeclaredFields("compactAssignment", "space", "reference", "compactId");
                return new PortableCompactAssignment(
                    ParseSpace(entry.RequireText("space")),
                    entry.RequireText("reference"),
                    new PortableCompactIdentifier(entry.RequireInt32("compactId")));
            }));
    }

    private static PortableIdentitySpace ParseSpace(string token) => token switch
    {
        "operation" => PortableIdentitySpace.Operation,
        "shape" => PortableIdentitySpace.Shape,
        "fragment" => PortableIdentitySpace.Fragment,
        _ => throw PortableFaultException.Malformed("compact-space", $"'{token}' is not a compact-identifier space.")
    };
}

/// <summary>The body of a <c>request</c> envelope.</summary>
/// <remarks>
/// The Operation may be named canonically or by a compact identifier this binding assigned. A
/// compact identifier the binding never assigned resolves to no canonical identity, so it is an
/// unsupported contract rather than an unknown Operation.
/// </remarks>
public sealed record PortableRequestBody(
    PortableOperationReference? Operation,
    int? CompactOperation,
    PortableShapeReference InputShape,
    CborItem Input,
    ImmutableArray<CborItem> Resources)
{
    private static readonly string[] Fields = ["operation", "compactOperation", "inputShape", "input", "resources"];

    public CborItem Encode()
    {
        var entries = new List<(string, CborItem)>
        {
            ("inputShape", CborAccess.Reference(InputShape.Name.Value, InputShape.Version)),
            ("input", Input),
            ("resources", new CborArray(Resources))
        };

        if (Operation is { } operation)
        {
            entries.Add(("operation", CborAccess.Reference(operation.Name.Value, operation.Version)));
        }

        if (CompactOperation is { } compact)
        {
            entries.Add(("compactOperation", new CborInteger(compact)));
        }

        return CborMap.Of(entries);
    }

    public static PortableRequestBody Decode(CborItem item)
    {
        var map = CborAccess.RequireMap(item, "request");
        map.RequireDeclaredFields("request", Fields);
        foreach (var entry in map.Entries.Where(entry => !string.Equals(entry.Key, "input", StringComparison.Ordinal)))
        {
            PortableForbiddenContent.RequireCleanControlItem(entry.Value);
        }

        PortableForbiddenContent.RequireCleanPayloadItem(map.Require("input"));

        var hasOperation = map.Contains("operation");
        var hasCompact = map.Contains("compactOperation");
        if (hasOperation == hasCompact)
        {
            throw PortableFaultException.Malformed(
                "operation-designation",
                "A request names its Operation either canonically or by one compact identifier.");
        }

        PortableOperationReference? operation = null;
        if (hasOperation)
        {
            var reference = CborAccess.ReadReference(map.Require("operation"), "operation");
            operation = new PortableOperationReference(reference.Name, reference.Version);
        }

        return new PortableRequestBody(
            operation,
            hasCompact ? map.RequireInt32("compactOperation") : null,
            PortableContractCodec.ReadShapeReference(map.Require("inputShape"), "inputShape"),
            map.Require("input"),
            [.. map.RequireArray("resources").Items]);
    }
}

public enum PortableOutcomeStatus
{
    Succeeded,
    Failed
}

/// <summary>The body of an <c>outcome</c> envelope.</summary>
public sealed record PortableOutcomeBody(
    PortableOutcomeStatus Status,
    PortableShapeReference ValueShape,
    CborItem Value,
    long ProviderEffectCount)
{
    private static readonly string[] Fields = ["status", "valueShape", "value", "providerEffectCount"];

    public CborItem Encode() => CborMap.Of(
    [
        ("status", new CborText(Status == PortableOutcomeStatus.Succeeded ? "succeeded" : "failed")),
        ("valueShape", CborAccess.Reference(ValueShape.Name.Value, ValueShape.Version)),
        ("value", Value),
        ("providerEffectCount", new CborInteger(ProviderEffectCount))
    ]);

    public static PortableOutcomeBody Decode(CborItem item)
    {
        var map = CborAccess.RequireMap(item, "outcome");
        map.RequireDeclaredFields("outcome", Fields);
        foreach (var entry in map.Entries.Where(entry => !string.Equals(entry.Key, "value", StringComparison.Ordinal)))
        {
            PortableForbiddenContent.RequireCleanControlItem(entry.Value);
        }

        PortableForbiddenContent.RequireCleanPayloadItem(map.Require("value"));

        var status = map.RequireText("status") switch
        {
            "succeeded" => PortableOutcomeStatus.Succeeded,
            "failed" => PortableOutcomeStatus.Failed,
            var other => throw PortableFaultException.Malformed(
                "outcome-status",
                $"'{other}' is not a terminal Outcome status.")
        };

        return new PortableOutcomeBody(
            status,
            PortableContractCodec.ReadShapeReference(map.Require("valueShape"), "valueShape"),
            map.Require("value"),
            map.RequireInteger("providerEffectCount"));
    }
}

/// <summary>The body of a <c>protocol-error</c> envelope.</summary>
/// <remarks>
/// The portable category drives semantics. The local code travels beside it as non-normative data,
/// so two realizations may use different local strings for the same category.
/// </remarks>
public sealed record PortableProtocolErrorBody(
    PortableProtocolCategory Category,
    string LocalCode,
    PortableFailureDomain FailureDomain)
{
    public CborItem Encode() => CborMap.Of(
    [
        ("category", new CborText(Category.Token())),
        ("localCode", new CborText(LocalCode)),
        ("failureDomain", new CborText(FailureDomain.Token()))
    ]);

    public static PortableProtocolErrorBody Decode(CborItem item)
    {
        var map = CborAccess.RequireMap(item, "protocol-error");
        map.RequireDeclaredFields("protocol-error", "category", "localCode", "failureDomain");
        PortableForbiddenContent.RequireCleanControlItem(map);
        return new PortableProtocolErrorBody(
            ParseCategory(map.RequireText("category")),
            map.RequireText("localCode"),
            ParseDomain(map.RequireText("failureDomain")));
    }

    private static PortableProtocolCategory ParseCategory(string token)
    {
        foreach (var category in Enum.GetValues<PortableProtocolCategory>())
        {
            if (string.Equals(category.Token(), token, StringComparison.Ordinal))
            {
                return category;
            }
        }

        throw PortableFaultException.Malformed("protocol-category", $"'{token}' is outside the Channel taxonomy.");
    }

    private static PortableFailureDomain ParseDomain(string token)
    {
        foreach (var domain in Enum.GetValues<PortableFailureDomain>())
        {
            if (string.Equals(domain.Token(), token, StringComparison.Ordinal))
            {
                return domain;
            }
        }

        throw PortableFaultException.Malformed("failure-domain", $"'{token}' is outside the Channel taxonomy.");
    }
}

File: Minimal/src/Brontide.Minimal.Binding/Portable/PortableEnvelope.fs
namespace Brontide.Minimal.Binding.Portable

/// One Channel envelope.
[<StructuralEquality; NoComparison>]
type Envelope =
    { ContractVersion: int
      Kind: EnvelopeKind
      Channel: ChannelId
      Request: ChannelRequestId option
      Execution: ChannelExecutionId option
      Body: CborItem }

[<RequireQualifiedAccess>]
module Envelope =

    let control kind channel body =
        { ContractVersion = ContractDocument.SupportedContractVersion
          Kind = kind
          Channel = channel
          Request = None
          Execution = None
          Body = body }

    let empty kind channel = control kind channel (CborMap [])

    let correlated kind channel request execution body =
        { ContractVersion = ContractDocument.SupportedContractVersion
          Kind = kind
          Channel = channel
          Request = request
          Execution = execution
          Body = body }

/// Strict envelope codec; every miss is a portable category rather than a parser failure.
[<RequireQualifiedAccess>]
module EnvelopeCodec =

    let private envelopeFields =
        [ "contractVersion"; "kind"; "channelId"; "requestId"; "executionId"; "body" ]

    let toItem (envelope: Envelope) =
        CborMap(
            [ "contractVersion", CborInteger(int64 envelope.ContractVersion)
              "kind", CborText(EnvelopeKind.token envelope.Kind)
              "channelId", CborText(ChannelId.value envelope.Channel)
              "body", envelope.Body ]
            @ (match envelope.Request with
               | Some request -> [ "requestId", CborText(ChannelRequestId.value request) ]
               | None -> [])
            @ (match envelope.Execution with
               | Some execution -> [ "executionId", CborText(ChannelExecutionId.value execution) ]
               | None -> [])
        )

    let encode envelope = PortableCbor.encode (toItem envelope)

    let ofItem (item: CborItem) : PortableResult<Envelope> =
        portable {
            let! entries = CborAccess.requireMap item "envelope"

            // Foreign runtime identity is refused before the field set is checked, so the rejection
            // names the rule that actually applies rather than reporting an unknown field.
            do!
                entries
                |> iterate (fun (key, value) ->
                    if key = "body" then
                        Ok()
                    elif PortableForbiddenContent.isForbidden key then
                        malformed "foreign-runtime-data" $"Control field '{key}' carries foreign runtime identity."
                    else
                        PortableForbiddenContent.requireCleanControl value)

            do! CborAccess.requireDeclaredFields entries "envelope" envelopeFields
            let! contractVersion = CborAccess.int32 entries "contractVersion"

            do!
                ensure (contractVersion = ContractDocument.SupportedContractVersion) (fun () ->
                    refuse
                        ProtocolCategory.UnsupportedVersion
                        "envelope-version"
                        $"Envelope contract version {contractVersion} is not recognized.")

            let! kindToken = CborAccess.text entries "kind"
            let! kind = EnvelopeKind.tryParse kindToken
            let! channel = CborAccess.text entries "channelId"
            let! requestText = CborAccess.optionalText entries "requestId"
            let! executionText = CborAccess.optionalText entries "executionId"

            do!
                ensure (not (EnvelopeKind.requiresRequestId kind) || requestText.IsSome) (fun () ->
                    malformed
                        "correlation-absent"
                        $"A '{EnvelopeKind.token kind}' envelope carries a request identity; it is never matched to an outstanding request by position.")

            let! body = CborAccess.field entries "body"

            return
                { ContractVersion = contractVersion
                  Kind = kind
                  Channel = ChannelId.received channel
                  Request = requestText |> Option.map ChannelRequestId.received
                  Execution = executionText |> Option.map ChannelExecutionId.received
                  Body = body }
        }

    let decode (frame: byte array) limits =
        PortableCbor.decode frame limits |> Result.bind ofItem

/// The body of an establish-accepted envelope.
type EstablishAcceptedBody =
    { Contract: ContractDocument
      CompactIdentifiers: CompactAssignment list }

[<RequireQualifiedAccess>]
module EstablishAcceptedBody =

    let encode (body: EstablishAcceptedBody) =
        CborMap
            [ "contract", ContractCodec.encode body.Contract
              "compactIdentifiers",
              CborArray(
                  body.CompactIdentifiers
                  |> List.map (fun assignment ->
                      CborMap
                          [ "space", CborText(IdentitySpace.token assignment.Space)
                            "reference", CborText assignment.Reference
                            "compactId", CborInteger(int64 (CompactId.value assignment.Compact)) ])
              ) ]

    let decode item : PortableResult<EstablishAcceptedBody> =
        portable {
            let! entries = CborAccess.requireMap item "establish-accepted"
            do! CborAccess.requireDeclaredFields entries "establish-accepted" [ "contract"; "compactIdentifiers" ]
            let! contract = CborAccess.field entries "contract" |> Result.bind ContractCodec.decode

            let! assignments =
                CborAccess.arrayOf entries "compactIdentifiers" (fun element ->
                    portable {
                        let! assignment = CborAccess.requireMap element "compactAssignment"
                        do! CborAccess.requireDeclaredFields assignment "compactAssignment" [ "space"; "reference"; "compactId" ]
                        let! spaceToken = CborAccess.text assignment "space"
                        let! space = IdentitySpace.tryParse spaceToken
                        let! reference = CborAccess.text assignment "reference"
                        let! compactValue = CborAccess.int32 assignment "compactId"
                        let! compact = CompactId.tryCreate compactValue

                        return
                            { Space = space
                              Reference = reference
                              Compact = compact }
                    })

            return
                { Contract = contract
                  CompactIdentifiers = assignments }
        }

/// How a request names the Operation it invokes.
///
/// The two designations are exclusive by construction rather than by a pair of optional members
/// that could both be present or both absent on the wire.
[<RequireQualifiedAccess>]
type OperationDesignation =
    | Canonical of PortableOperationRef
    /// A compact identifier this binding assigned. One it never assigned resolves to no canonical
    /// identity, which is an unsupported contract rather than an unknown Operation.
    | Compact of int

/// The body of a request envelope.
type RequestBody =
    { Operation: OperationDesignation
      InputShape: PortableShapeRef
      Input: CborItem
      Resources: CborItem list }

[<RequireQualifiedAccess>]
module RequestBody =

    let private fields = [ "operation"; "compactOperation"; "inputShape"; "input"; "resources" ]

    let encode (body: RequestBody) =
        CborMap(
            [ "inputShape", CborAccess.encodeCanonical (PortableShapeRef.canonical body.InputShape)
              "input", body.Input
              "resources", CborArray body.Resources ]
            @ (match body.Operation with
               | OperationDesignation.Canonical reference ->
                   [ "operation", CborAccess.encodeCanonical (PortableOperationRef.canonical reference) ]
               | OperationDesignation.Compact compact -> [ "compactOperation", CborInteger(int64 compact) ])
        )

    let decode item : PortableResult<RequestBody> =
        portable {
            let! entries = CborAccess.requireMap item "request"
            do! CborAccess.requireDeclaredFields entries "request" fields

            do!
                entries
                |> iterate (fun (key, value) ->
                    if key = "input" then
                        Ok()
                    else
                        PortableForbiddenContent.requireCleanControl value)

            let! input = CborAccess.field entries "input"
            do! PortableForbiddenContent.requireCleanPayload input

            let hasOperation = CborAccess.contains entries "operation"
            let hasCompact = CborAccess.contains entries "compactOperation"

            do!
                ensure (hasOperation <> hasCompact) (fun () ->
                    malformed
                        "operation-designation"
                        "A request names its Operation either canonically or by one compact identifier.")

            let! designation =
                if hasOperation then
                    CborAccess.field entries "operation"
                    |> Result.bind (fun item -> CborAccess.readCanonical item "operation")
                    |> Result.map (PortableOperationRef.ofCanonical >> OperationDesignation.Canonical)
                else
                    CborAccess.int32 entries "compactOperation" |> Result.map OperationDesignation.Compact

            let! inputShape =
                CborAccess.field entries "inputShape"
                |> Result.bind (fun item -> CborAccess.readCanonical item "inputShape")
                |> Result.map PortableShapeRef.ofCanonical

            let! resources = CborAccess.array entries "resources"

            return
                { Operation = designation
                  InputShape = inputShape
                  Input = input
                  Resources = resources }
        }

[<RequireQualifiedAccess>]
type OutcomeStatus =
    | Succeeded
    | Failed

/// The body of an outcome envelope.
type OutcomeBody =
    { Status: OutcomeStatus
      ValueShape: PortableShapeRef
      Value: CborItem
      ProviderEffectCount: int64 }

[<RequireQualifiedAccess>]
module OutcomeBody =

    let private fields = [ "status"; "valueShape"; "value"; "providerEffectCount" ]

    let encode (body: OutcomeBody) =
        CborMap
            [ "status",
              CborText(
                  match body.Status with
                  | OutcomeStatus.Succeeded -> "succeeded"
                  | OutcomeStatus.Failed -> "failed"
              )
              "valueShape", CborAccess.encodeCanonical (PortableShapeRef.canonical body.ValueShape)
              "value", body.Value
              "providerEffectCount", CborInteger body.ProviderEffectCount ]

    let decode item : PortableResult<OutcomeBody> =
        portable {
            let! entries = CborAccess.requireMap item "outcome"
            do! CborAccess.requireDeclaredFields entries "outcome" fields

            do!
                entries
                |> iterate (fun (key, value) ->
                    if key = "value" then
                        Ok()
                    else
                        PortableForbiddenContent.requireCleanControl value)

            let! value = CborAccess.field entries "value"
            do! PortableForbiddenContent.requireCleanPayload value
            let! statusToken = CborAccess.text entries "status"

            let! status =
                match statusToken with
                | "succeeded" -> Ok OutcomeStatus.Succeeded
                | "failed" -> Ok OutcomeStatus.Failed
                | other -> malformed "outcome-status" $"'{other}' is not a terminal Outcome status."

            let! valueShape =
                CborAccess.field entries "valueShape"
                |> Result.bind (fun item -> CborAccess.readCanonical item "valueShape")
                |> Result.map PortableShapeRef.ofCanonical

            let! providerEffectCount = CborAccess.integer entries "providerEffectCount"

            return
                { Status = status
                  ValueShape = valueShape
                  Value = value
                  ProviderEffectCount = providerEffectCount }
        }

/// The body of a protocol-error envelope.
///
/// The portable category drives semantics. The local code travels beside it as non-normative data,
/// so two realizations may use different local strings for the same category.
type ProtocolErrorBody =
    { Category: ProtocolCategory
      LocalCode: string
      FailureDomain: FailureDomain }

[<RequireQualifiedAccess>]
module ProtocolErrorBody =

    let encode (body: ProtocolErrorBody) =
        CborMap
            [ "category", CborText(ProtocolCategory.token body.Category)
              "localCode", CborText body.LocalCode
              "failureDomain", CborText(FailureDomain.token body.FailureDomain) ]

    let decode item : PortableResult<ProtocolErrorBody> =
        portable {
            let! entries = CborAccess.requireMap item "protocol-error"
            do! CborAccess.requireDeclaredFields entries "protocol-error" [ "category"; "localCode"; "failureDomain" ]
            do! PortableForbiddenContent.requireCleanControl (CborMap entries)
            let! categoryToken = CborAccess.text entries "category"

            let! category =
                match ProtocolCategory.tryParse categoryToken with
                | Some category -> Ok category
                | None -> malformed "protocol-category" $"'{categoryToken}' is outside the Channel taxonomy."

            let! localCode = CborAccess.text entries "localCode"
            let! domainToken = CborAccess.text entries "failureDomain"

            let! domain =
                match FailureDomain.tryParse domainToken with
                | Some domain -> Ok domain
                | None -> malformed "failure-domain" $"'{domainToken}' is outside the Channel taxonomy."

            return
                { Category = category
                  LocalCode = localCode
                  FailureDomain = domain }
        }
